### PR TITLE
Add smp v2 notebook examples.

### PR DESCRIPTION
**Pull request by sheilaliuxl**
*Issue #, if available:*

- N.A.

*Description of changes:*

- Adding SMP v2 examples, to release in Dec 2023.

*Testing done:*

Run E2E tests with the example notebook, and then cleanup
1. Team specific AWS account setup and
2. All outputs.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [x] I have linted my notebook(s) and code using `black-nb -l 100 {path}/{notebook-name}.ipynb`

```
sliuxl@c889f3b957d9 15:49 ~/amazon-sagemaker-examples/training/distributed_training/pytorch/model_parallel_v2/gpt-neox $ black-nb -l 100  smp-train-gpt-neox-fsdp-tp.ipynb 
    25 cells left unchanged.
All done! ✨ 🍰 ✨
1 file left unchanged.


sliuxl@c889f3b957d9 15:49 ~/amazon-sagemaker-examples/training/distributed_training/pytorch/model_parallel_v2/gpt-neox $ git status smp-train-gpt-neox-fsdp-tp.ipynb 
On branch smp-v2-notebook
nothing to commit, working tree clean
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb [hunk 1]
```diff
@@ -0,0 +1,1032 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# @title Image URI.\n",
+    "\n",
+    "image = (\n",
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb:830]: I think we should not specify image, instead we should only specify `framework_version=2.0.1` in the `Estimator` and allow our changes to Python SDK to handle appropriate parsing of image URI based on instance type.

Comment (yongyanrao) [reply training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb:830]: Yes, just tested that `py_version="py310"` and `framework_version="2.0.1"` cannot pull the correct image.

Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb:830]: I believe our SDK changes havent been merged yet. I'd tested locally and it had worked

File: training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb [hunk 2]
```diff
@@ -0,0 +1,959 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# INTERNAL USE ONLY\n",
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb:7]: Remove this

File: training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb (hunk 1 above)
Comment (sheilaliuxl) [reply training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb:830]: WIP for me now, testing with a SM notebook.

File: training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb [hunk 3]
```diff
@@ -0,0 +1,959 @@
...
+    "**NOTE:** This step might require a kernel restart."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "CONTAINER_IMAGE = \"...\"\n",
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb:64]: I think we should specify why these variables are required, or add them to the relevant cells, e.g fsx creation cell for fsx variables

File: training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb [hunk 4]
```diff
@@ -0,0 +1,959 @@
...
+    "    data_channels = {\"train\": train, \"test\": train}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Set hyperparameters, metric definitions, and MPI options\n",
+    "#### Tensor Parallelism\n",
+    "TODO(viczhu):\n",
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb:585]: Remove

File: training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb [hunk 5]
```diff
@@ -0,0 +1,959 @@
...
+    "                    \"tensor_parallel_degree\": tensor_parallel_degree,\n",
+    "                    \"hybrid_shard_degree\": hybrid_shard_degree,\n",
+    "                    \"sm_activation_offloading\": offload_activations,\n",
+    "                    \"activation_loading_horizon\": activation_loading_horizon,\n",
+    "                },\n",
+    "            }\n",
+    "        },\n",
+    "    },\n",
+    "    py_version=\"py310\",\n",
+    "    image_uri=image,\n",
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb:875]: Please update to use `framework_version` instead

File: training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb [hunk 6]
```diff
@@ -0,0 +1,971 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# INTERNAL USE ONLY\n",
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb:7]: Remove

File: training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb [hunk 7]
```diff
@@ -0,0 +1,971 @@
...
+    "    data_channels = {\"train\": train, \"test\": train}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Set hyperparameters, metric definitions, and MPI options\n",
+    "#### Tensor Parallelism\n",
+    "TODO(viczhu):\n",
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb:592]: Remove

File: training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb (hunk 3 above)
Comment (sheilaliuxl) [reply training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb:64]: It's intentional to consolidate those vars, as internal vs external values are different.
- Easier for cleanup for the github repo.

File: training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb (hunk 2 above)
Comment (yongyanrao) [reply training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb]: Updates

File: training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb (hunk 4 above)
Comment (yongyanrao) [reply training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb]: Updates

File: training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb (hunk 5 above)
Comment (yongyanrao) [reply training/distributed_training/pytorch/model_parallel_v2/llama_v2/smp-finetuning-llama-fsdp-tp.ipynb]: Updated

File: training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb (hunk 6 above)
Comment (yongyanrao) [reply training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb]: Updated

File: training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb (hunk 7 above)
Comment (yongyanrao) [reply training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb]: Updated

File: training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb (hunk 1 above)
Comment (yongyanrao) [reply training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb:830]: Was able to test it successfully. Will update it.

Comment (sheilaliuxl) [reply training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-train-gpt-neox-fsdp-tp.ipynb]: Fixed as suggested, PTAL.

File: training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb [hunk 8]
```diff
@@ -0,0 +1,969 @@
...
+    "if not use_fsx:\n",
+    "    # If you want to resume training, set checkpoint_s3_uri to the same path as a previous job.\n",
+    "    # Previous checkpoint to load must have same model config.\n",
+    "    checkpoint_bucket = f\"s3://sagemaker-{region}-{account}/\"\n",
+    "    checkpoint_s3_uri = (\n",
+    "        f\"{checkpoint_bucket}/experiments/smp_fsdp-{model_type}-checkpoints/{base_job_name}/\"\n",
+    "    )"
+   ]
+  },
+  {
```
Comment (sheilaliuxl) [inline training/distributed_training/pytorch/model_parallel_v2/gpt-neox/smp-finetuning-gpt-neox-fsdp-tp.ipynb:835]: nit: Ideally we can remove the whole cell here, and def of `CONTAINER_IMAGE`.

**Review by SuhitK — APPROVED** [review]
LGTM

**Review by akrishna1995 — APPROVED** [review]

**Pull request merged**